Port to nusb

Dirbaio opened the pull request:
[nusb](https://github.com/kevinmehall/nusb) is a pure-Rust libusb replacement. It can help a lot with the issues we're having with libusb (cross compiling pain, segfaults).

- [x] Port probe drivers
  - [x] stlink
  - [x] cmsisdap
  - [x] espusbjtag
  - [x] jaylink
  - [x] ftdi
  - [x] wlink
- [x] Test probe drivers
  - [x] stlink
  - [x] cmsisdap
  - [x] espusbjtag
  - [x] jaylink
  - [x] ftdi
  - [ ] wlink
- [x] Tested working on Linux
- [x] Tested working on Windows
- [x] nusb macos support support  https://github.com/kevinmehall/nusb/issues/3
- [ ] test on macos
- [x] ~~jaylink release~~ integrated jaylink into probe-rs.

~~I'm not sure what strategy we should follow. Either port probe drivers one-by-one, or port them all. Also, nusb doesn't support macos yet...~~

Tiwalun commented:
Nice work! I've started working on this for cmsis-dap. Unfortunately there is still some functionality missing in nusb for that, e.g. related to interfaces. 

Maybe we could put it behind a feature? Then we can merge the PRs, but still keep on using libusb until we can fully switch. I'm not sure if we gain anything by switching only some parts.

Dirbaio commented:
a feature sounds nice yes. downside is we have to maintain both versions of the code, but in the case of stlink at least it's not that much code.

Dirbaio commented:
So... I went ahead and ported all probes. Seems to work fine!

I'm not looking forward to add features for all probe drivers and for nusb/rusb. I think I'd rather leave this as an experimental branch I'll rebase periodically, and wait until macos nusb is ready.

Tiwalun commented:
Could you please replace the relative paths in the `Cargo.toml` with git dependencies? A bit hard to build otherwise :)

Dirbaio commented:
@Tiwalun done

Yatekii commented:
Hmm why is espusbjtag not supported? Is this not ported to the shared JTAG driver?

Dirbaio commented:
@Yatekii I've ported it, I just didn't test it.

Yatekii commented:
Ah nice, I took the checkmark as not done :D Thanks!

MabezDev commented:
There is a trace log that panics with esp_usb_jtag: https://github.com/probe-rs/probe-rs/pull/1842/files#diff-9aec9621527157e6daf555546fa14a8bc22c0e088b3ff5eb058280bf4c78d110R183.

With that removed I can confirm esp_usb_jtag works with nusb :tada:. It is a little bit slower though, probe-rs master can flash an esp32c3 at 30KB/s, with this pr its more like 24KB/s, not a huge deal but I thought I'd point it out.

kevinmehall commented:
@MabezDev what OS was this on?
> It is a little bit slower though, probe-rs master can flash an esp32c3 at 30KB/s, with this pr its more like 24KB/s, not a huge deal but I thought I'd point it out.

While the sync API in libusb is also built on async at the OS level, one difference is that libusb runs the `poll` loop from within the blocking call on Linux and Windows, while nusb runs it on a dedicated thread, and notifies `block_on` via a waker. That might add a bit of latency for single-threaded synchronous use, but I don't think it would be that significant.

Also confirm you compiled with `--release`? Having it pure Rust means it's all affected by your Rust optimization level.

MabezDev commented:
> @MabezDev what OS was this on?

Linux, kernel 6.1.

> Also confirm you compiled with --release? Having it pure Rust means it's all affected by your Rust optimization level.

Correct, I'm installing with cargo install, which builds it in release mode.

It would be good to see if it affects any other probes, or if this is a specific espusbjtag issue. @Dirbaio Would you mind comparing stlink or pico probe perf?

Yatekii commented:
Wheee mac support was merged 🎉

Dirbaio commented:
I see no perf difference with stlink nor cmsisdap on linux.

@bugadani just retested espusbjtag and found it's within 1%, seems the bottleneck was trace logging to file being enabled by default (nusb is more verbose maybe).

jonathanpallant commented:
```console
➜  probe-rs git:(nusb) uname -a
Darwin MacBook-Pro 23.1.0 Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:24 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6000 arm64
➜  probe-rs git:(nusb) cargo run --release --features=cli --bin probe-rs list
    Finished release [optimized] target(s) in 0.15s
     Running `target/release/probe-rs list`
The following debug probes were found:
[0]: STLink V2-1 (VID: 0483, PID: 374b, Serial: 066DFF495177514867214535, StLink)
➜  probe-rs git:(nusb) cargo run --release --features=cli --bin probe-rs info
    Finished release [optimized] target(s) in 0.24s
     Running `target/release/probe-rs info`
Probing target via JTAG

ARM Chip:
Debug Port: Version 1, DP Designer: ARM Ltd

Unable to debug RISC-V targets using the current probe. RISC-V specific information cannot be printed.
Unable to debug Xtensa targets using the current probe. Xtensa specific information cannot be printed.

Probing target via SWD

ARM Chip:
Debug Port: Version 1, DP Designer: ARM Ltd

Unable to debug RISC-V targets using the current probe. RISC-V specific information cannot be printed.
Unable to debug Xtensa targets using the current probe. Xtensa specific information cannot be printed.
```

jonathanpallant commented:
OK, I found a thing I can actually flash (the STM32 Nucleo won't let me connect - even using regular probe-rs; weird)...

```console
➜  radio-app git:(main) /Users/jonathan/Documents/probe-rs/target/release/probe-rs run --chip nRF52840_xxAA target/thumbv7em-none-eabihf/debug/blinky
      Erasing ✔ [00:00:00] [########################################################] 8.00 KiB/8.00 KiB @ 32.71 KiB/s (eta 0s )
  Programming ✔ [00:00:00] [########################################################] 8.00 KiB/8.00 KiB @ 40.62 KiB/s (eta 0s )    Finished in 0.46s
LED toggled at Duration { secs: 0, nanos: 999969481 }
LED toggled at Duration { secs: 2, nanos: 30516 }
LED toggled at Duration { secs: 3, nanos: 91551 }
LED toggled at Duration { secs: 4, nanos: 152587 }
LED toggled at Duration { secs: 5, nanos: 183103 }
LED toggled at Duration { secs: 6, nanos: 244140 }
```

jonathanpallant commented:
One last macOS check for you:

```console
➜  radio-app git:(main) otool -L /Users/jonathan/Documents/probe-rs/target/release/probe-rs
/Users/jonathan/Documents/probe-rs/target/release/probe-rs:
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 2202.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2487.30.104)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)

➜  radio-app git:(main) which probe-rs
/Users/jonathan/.cargo/bin/probe-rs

➜  radio-app git:(main) otool -L /Users/jonathan/.cargo/bin/probe-rs
/Users/jonathan/.cargo/bin/probe-rs:
	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration (compatibility version 1.0.0, current version 1296.0.1)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 61040.1.3)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 2048.1.255)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/opt/homebrew/opt/libusb/lib/libusb-1.0.0.dylib (compatibility version 4.0.0, current version 4.0.0)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 2483.0.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.0.0)
➜  radio-app git:(main)
```

libusb has gone away, as expected.

Dirbaio commented:
@thejpster what probe is [this](https://github.com/probe-rs/probe-rs/pull/1842#issuecomment-1881522570) with? jlink

jonathanpallant commented:
`--chip nRF52840_xxAA` is me using an nRF52840-DK, so yeah, that's a J-Link.

The STM32 Nucleo has an ST-Link on it. I think OxidOS has locked my flash or something.

bugadani commented:
> `--chip nRF52840_xxAA` is me using an nRF52840-DK, so yeah, that's a J-Link.
> 
> The STM32 Nucleo has an ST-Link on it. I think OxidOS has locked my flash or something.
> 
> ```
> ➜  radio-app git:(main) /Users/jonathan/Documents/probe-rs/target/release/probe-rs list
> The following debug probes were found:
> [0]: J-Link (J_Link) (VID: 1366, PID: 1015, Serial: 000683932144, JLink)
> ```

Just to be clear, during your tests only one probe was connected, so it's expected behaviour for only one to show up in `probe-rs list`, right?

bugadani commented on probe-rs/src/probe/stlink/usb_interface.rs:
```diff
@@ -100,23 +100,17 @@ pub(crate) trait StLinkUsb: std::fmt::Debug {
                     // If the VID & PID match, match the serial if one was given.
-                    if let Some(serial) = selector.serial_number.as_ref() {
-                        let sn_str = read_serial_number(&device, &descriptor).ok();
+                    if let Some(serial) = &selector.serial_number {
```
Why do you prefer `&` when we already have `.as_ref()` two lines down?

Dirbaio replied:
no reason at all, changed it back to `.as_ref()`

Dirbaio commented:
- Rebased
- Removed the "TODO: this sometimes randomly fails. Race condition reading descriptors maybe" and uncommented the code. I can't get it to fail anymore, I'm not sure what was wrong :shrug: 
- Addressed review comment.

bugadani reviewed (approved):
Thank you!

noppej commented:
@Yatekii Does this imply the prerequisites in the probe.rs docs can be removed?

Yatekii commented:
Oh yes, that's correct :)

Tiwalun commented:
We should wait until this is actually released.

Yatekii commented:
Ah, sorry, I was mostly thinking of the README, not the webpage yet. My bad.

thejpster commented:
> Just to be clear, during your tests only one probe was connected, so it's expected behaviour for only one to show up in `probe-rs list`, right?

Yes, I unplugged the Nucleo and used the same cable for the nRF-DK.
